test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -293,54 +225,94 @@ static subgroups::cl_half to_half(float x)
     return value;
 }
 
-template <> struct OPERATION<subgroups::cl_half, 0>
+// for integer types
+template <typename Ty> inline Ty calculate(Ty a, Ty b, ArithmeticOp operation)
 {
-    static subgroups::cl_half calculate(subgroups::cl_half a,
-                                        subgroups::cl_half b)
+    switch (operation)
     {
-        return to_half(to_float(a) + to_float(b));
+        case ArithmeticOp::add_: return a + b;
+        case ArithmeticOp::max_: return a > b ? a : b;
+        case ArithmeticOp::min_: return a < b ? a : b;
+        case ArithmeticOp::mul_: return a * b;
+        case ArithmeticOp::and_: return a & b;
+        case ArithmeticOp::or_: return a | b;
+        case ArithmeticOp::xor_: return a ^ b;
+        case ArithmeticOp::logical_and: return a && b;
+        case ArithmeticOp::logical_or: return a || b;
+        case ArithmeticOp::logical_xor: return !a ^ !b;
+        default: log_error("Unknown operation request"); break;
     }
-};
-
-template <> struct OPERATION<subgroups::cl_half, 1>
+    return 0;
+}
+// speciallize for floating points.
+template <>
+inline cl_double calculate(cl_double a, cl_double b, ArithmeticOp operation)
```

Comment:
I'm not sure the inline attribute on templates makes any difference.

Reply:
I got compiler errors on Linux while Linking.